src/Lean/Data/Lsp/PositionEncoding.lean

```diff
@@ -0,0 +1,140 @@
+/-
+Copyright (c) 2020 Marc Huisinga. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Authors: Marc Huisinga, Wojciech Nawrocki
+-/
+import Init.Data.String
+import Init.Data.Array
+import Lean.Data.Lsp.Basic
+import Lean.Data.Lsp.PositionEncodingKind
+import Lean.Data.Position
+
+
+/-! LSP uses UTF-16 for indexing, so we need to provide some primitives
+to interact with Lean strings using UTF-16 indices. -/
+
+namespace Char
+
+def utf16Size (c : Char) : UInt32 :=
+  if c.val ≤ 0xFFFF then 1 else 2
```

Comment:
Bikeshed, but is `Lean.Data.Lsp.PositionEncoding` the right name for this? The module contains a bunch of general UTF utilities that are not about LSP positions. Maybe these could be moved to something like `Lean.Data.Utf`?

Reply:
I think that the underlying principle would be "When the compiler needs a thing that seems useful, make it available in a plausible module that people can import." I don't really believe this principle - I think the result will be either a compiler that contains too many things that it doesn't really need, or an incomplete API that's full of holes for other uses.

It could be that we need a relatively complete Unicode encodings API - then we might as well do it, as we've done for e.g. lists and strings. But so far, we seem to just need a little bit, so I'd rather keep them next to the code that uses them. Perhaps it should be marked `private` though :-)

Reply:
(also, LSP positions are all about UTF encodings, due to the decision of the protocol designers to mandate UTF-8 for messages, and UTF-16 for source positions - a correct LSP implementation needs lots of this kind of thing)

Reply:
That's fair enough! I wouldn't `private` things though, historically that mostly leads to annoyed people copy-pasting code into their own repository.